### ISSUE TITLE
Investigate API stability

### ISSUE THREAD

**Issue by szilardcserey** [canonical-web-and-design/ubuntu-com-security-api#59]
CVE API query [ https://ubuntu.com/security/cves.json ] on package "linux" with limit 20 or without any limit results in "504 Gateway Time-out" , it only works with limit of 10 or less than 10

## Summary

CVE API query [ https://ubuntu.com/security/cves.json ]
on package "linux" with limit 20 or without any limit results in "504 Gateway Time-out" , it only works with limit of 10 or less than 10

## Process

curl -s --connect-timeout 600 -m 600 -X GET -H "Content-Type: application/json" "https://ubuntu.com/security/cves.json?package=linux&limit=20"
504 Gateway Time-out
The server didn't respond in time.

curl -s --connect-timeout 600 -m 600 -X GET -H "Content-Type: application/json" "https://ubuntu.com/security/cves.json?package=linux&limit=10" | python -m json.tool | grep "^ \"id"
"id": "CVE-2021-0920",
"id": "CVE-2021-3736",
"id": "CVE-2021-43389",
"id": "CVE-2020-27820",
"id": "CVE-2021-34981",
"id": "CVE-2021-43267",
"id": "CVE-2021-3892",
"id": "CVE-2021-43057",
"id": "CVE-2021-43056",
"id": "CVE-2021-3760",

## Current and expected result

Here I expected to get a list of 20 CVEs, the latest ones, but instead I got "504 Gateway Time-out"
curl -s --connect-timeout 600 -m 600 -X GET -H "Content-Type: application/json" "https://ubuntu.com/security/cves.json?package=linux&limit=20"
504 Gateway Time-out
The server didn't respond in time.

**Comment by mtruj013** [comment]
@szilardcserey  We've allotted time to work on this in the next iteration (about 2 weeks from now)

**Comment by szilardcserey** [comment]
Hi,

This issue is still not solved

The first 20 CVEs are correctly listed:

curl -s -X GET -H "Content-Type: application/json" "https://ubuntu.com/security/cves.json?package=linux-hwe-5.4&status=released&version=bionic&limit=20&offset=1" | python -m json.tool | egrep '^ {12}"id'
            "id": "CVE-2022-21499",
            "id": "CVE-2022-1116",
            "id": "CVE-2022-29581",
            "id": "CVE-2022-30594",
            "id": "CVE-2022-20008",
            "id": "CVE-2022-1516",
Hi,

This issue is still not solved

The first 20 CVEs are correctly listed:

curl -s -X GET -H "Content-Type: application/json" "https://ubuntu.com/security/cves.json?package=linux-hwe-5.4&status=released&version=bionic&limit=20&offset=1" | python -m json.tool | egrep '^ {12}"id'
            "id": "CVE-2022-21499",
            "id": "CVE-2022-1116",
            "id": "CVE-2022-29581",
            "id": "CVE-2022-30594",
            "id": "CVE-2022-20008",
            "id": "CVE-2022-1516",
            "id": "CVE-2022-1353",
            "id": "CVE-2021-0707",
            "id": "CVE-2022-1158",
            "id": "CVE-2022-28389",
            "id": "CVE-2022-28390",
            "id": "CVE-2022-28356",
            "id": "CVE-2022-1198",
            "id": "CVE-2022-1055",
            "id": "CVE-2022-1016",
            "id": "CVE-2021-4157",
            "id": "CVE-2022-0322",
            "id": "CVE-2021-4203",
            "id": "CVE-2021-4149",
            "id": "CVE-2022-27666",


However, querying the next 20 CVEs results in "504 Gateway Time-out"

curl -s -X GET -H "Content-Type: application/json" "https://ubuntu.com/security/cves.json?package=linux-hwe-5.4&status=released&version=bionic&limit=20&offset=20"
<html><body><h1>504 Gateway Time-out</h1>
The server didn't respond in time.
</body></html>
            "id": "CVE-2022-1353",
            "id": "CVE-2021-0707",
            "id": "CVE-2022-1158",
            "id": "CVE-2022-28389",
            "id": "CVE-2022-28390",
            "id": "CVE-2022-28356",
            "id": "CVE-2022-1198",
            "id": "CVE-2022-1055",
            "id": "CVE-2022-1016",
            "id": "CVE-2021-4157",
            "id": "CVE-2022-0322",
            "id": "CVE-2021-4203",
            "id": "CVE-2021-4149",
            "id": "CVE-2022-27666",


However, querying the next 20 CVEs results in "504 Gateway Time-out"

curl -s -X GET -H "Content-Type: application/json" "https://ubuntu.com/security/cves.json?package=linux-hwe-5.4&status=released&version=bionic&limit=20&offset=20"
<html><body><h1>504 Gateway Time-out</h1>
The server didn't respond in time.
</body></html>

**Comment by casual-lemon** [comment]
I asked yesterday about the curling the CVE's search functionality and it's under DDoS lately, so it's possible the service was failing when attempting to curl the API.

**Comment by casual-lemon** [comment]
I did some additonal testing and 
executed these commands

**Comment by carkod** [comment]
> I asked yesterday about the curling the CVE's search functionality and it's under DDoS lately, so it's possible the service was failing when attempting to curl the API.

So I'm so familiar with how cve works, but the DDoS is on search, but i think you are not using search you are using the cve REST endpoints. Search would be e.g. https://ubuntu.com/security/cves?q=application

**Comment by casual-lemon** [comment]
![gatewaytimeout2022-6-16 15-59](https://user-images.githubusercontent.com/62671519/174827410-8b492529-670b-46c2-a842-6db05a0355c6.png)
this was taken last week

**Comment by albertkol** [comment]
The issue seems to have been closed automatically on my PR merge. A bit of context...

@casual-lemon @szilardcserey 

I pushed a PR wish should help the API a bit. If the change is successful, I hope it will reduce the number of 504's we get.

Please let me know if you still experience them. In theory they can still happen, due to the API not being to handle the amount of requests it gets, but hopefully, less frequent.

But we no longer allow setting `limit` over `100`.

**Comment by szilardcserey** [comment]
Hi @albertkol 
Would it be possible to completely eradicate 504's

My customer said that they tried:

1.) 

https://ubuntu.com/security/cves?package=linux-hwe-5.4&status=released&version=bionic&limit=20&offset=140 

and still got 504 Gateway Time-out. 



2.) After that they tried linux package with various offsets: 

https://ubuntu.com/security/cves?package=linux-hwe-5.4&status=released&version=bionic&limit=20&offset=
xxx. 

a.) With xxx=50 and 75 it was OK

b.) offset=100 gave them Content Encoding Error

c.) offset=140 led to 504 Gateway Timeout.

**Comment by szilardcserey** [comment]
this results in 504 Gateway Timeout for me as well (tested with Chrome, Firefox and Brave browsers)

https://ubuntu.com/security/cves?package=linux-hwe-5.4&status=released&version=bionic&limit=20&offset=140

**Comment by nottrobin** [comment]
@albertkol I wonder if we could rename this issue to describe the actual problem that you fixed in your PR, and then open a new issue regarding @szilardcserey's ongoing issues?

**Comment by szilardcserey** [comment]
@albertkol @nottrobin interestingly this works for me 
https://ubuntu.com/security/cves?package=linux-hwe-5.4&status=released&version=bionic&limit=20&offset=100

it will display page 6

but when I click on the page buttons of page 7 or 8 or 10  at the bottom of the webpage, then might work, sometimes, but most of the time it ends up in " 504 Gateway Timeout " again

**Comment by szilardcserey** [comment]
Querying with cURL works fine no timeouts observed:




curl -s -X GET -H "Content-Type: application/json" "https://ubuntu.com/security/cves.json?package=linux-hwe-5.4&status=released&version=bionic&limit=20&offset=100" | python -m json.tool | egrep '^ {12}"id'
            "id": "CVE-2021-3752",
            "id": "CVE-2021-3764",
            "id": "CVE-2021-3744",
            "id": "CVE-2021-3772",
            "id": "CVE-2021-40490",
            "id": "CVE-2021-3759",
            "id": "CVE-2021-3753",
            "id": "CVE-2021-3739",
            "id": "CVE-2021-3732",
            "id": "CVE-2021-3743",
            "id": "CVE-2021-21781",
            "id": "CVE-2021-3656",
            "id": "CVE-2021-3653",
            "id": "CVE-2021-38208",
            "id": "CVE-2021-38207",
            "id": "CVE-2021-38205",
            "id": "CVE-2021-38204",
            "id": "CVE-2021-38199",
            "id": "CVE-2021-38198",
            "id": "CVE-2021-38160",

curl -s -X GET -H "Content-Type: application/json" "https://ubuntu.com/security/cves.json?package=linux-hwe-5.4&status=released&version=bionic&limit=20&offset=120" | python -m json.tool | egrep '^ {12}"id'
            "id": "CVE-2021-3655",
            "id": "CVE-2021-3679",
            "id": "CVE-2021-34556",
            "id": "CVE-2021-35477",
            "id": "CVE-2021-3659",
            "id": "CVE-2021-37576",
            "id": "CVE-2021-37159",
            "id": "CVE-2021-33909",
            "id": "CVE-2021-3640",
            "id": "CVE-2021-3612",
            "id": "CVE-2021-22555",
            "id": "CVE-2020-28097",
            "id": "CVE-2021-33624",
            "id": "CVE-2021-0605",
            "id": "CVE-2021-3600",
            "id": "CVE-2021-0512",
            "id": "CVE-2021-3609",
            "id": "CVE-2021-34693",
            "id": "CVE-2021-0129",
            "id": "CVE-2020-26558",


curl -s -X GET -H "Content-Type: application/json" "https://ubuntu.com/security/cves.json?package=linux-hwe-5.4&status=released&version=bionic&limit=20&offset=140" | python -m json.tool | egrep '^ {12}"id'
            "id": "CVE-2021-3564",
            "id": "CVE-2020-36386",
            "id": "CVE-2020-36385",
            "id": "CVE-2021-3573",
            "id": "CVE-2021-22543",
            "id": "CVE-2020-26555",
            "id": "CVE-2021-33034",
            "id": "CVE-2021-33033",
            "id": "CVE-2021-23134",
            "id": "CVE-2020-26147",
            "id": "CVE-2020-26145",
            "id": "CVE-2020-26141",
            "id": "CVE-2020-26139",
            "id": "CVE-2020-24588",
            "id": "CVE-2020-24587",
            "id": "CVE-2020-24586",
            "id": "CVE-2021-32399",
            "id": "CVE-2021-31829",
            "id": "CVE-2021-31916",
            "id": "CVE-2021-23133",

curl -s -X GET -H "Content-Type: application/json" "https://ubuntu.com/security/cves.json?package=linux-hwe-5.4&status=released&version=bionic&limit=20&offset=160" | python -m json.tool | egrep '^ {12}"id'
            "id": "CVE-2021-29155",
            "id": "CVE-2021-3506",
            "id": "CVE-2021-3493",
            "id": "CVE-2021-3492",
            "id": "CVE-2020-36322",
            "id": "CVE-2021-29154",
            "id": "CVE-2020-36312",
            "id": "CVE-2020-36311",
            "id": "CVE-2021-3483",
            "id": "CVE-2021-28688",
            "id": "CVE-2021-30002",
            "id": "CVE-2021-29650",
            "id": "CVE-2021-29647",
            "id": "CVE-2021-29265",
            "id": "CVE-2021-29264",
            "id": "CVE-2021-20292",
            "id": "CVE-2021-3444",
            "id": "CVE-2021-28972",
            "id": "CVE-2021-28971",
            "id": "CVE-2021-28964",

**Comment by szilardcserey** [comment]
Hi @albertkol  @carkod   @nottrobin
Would it be possible to completely eradicate  the "504 Gateway Timeout" when querying the API using browsers (Chrome, Firefox)?
or it was intentionally added because of DDoS and we have to live with it?

**Comment by mtruj013** [comment]
@szilardcserey I'm working on a PR that may accomplish that, at least for the timeout's you're getting for that query. I tested it locally and it worked, but it's still in review pending an unrelated bug fix. Once that is merged I'll ping you so we can test if the issue has persisted.
